src/drivers/actuators/vertiq_io/vertiq_telemetry_manager.cpp

```diff
@@ -0,0 +1,185 @@
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+#include "vertiq_telemetry_manager.hpp"
+
+VertiqTelemetryManager::VertiqTelemetryManager(IFCI *motor_interface) :
+	_motor_interface(motor_interface)
+{}
+
+void VertiqTelemetryManager::Init(uint16_t telem_bitmask)
+{
+	//On init, make sure to set our bitmask, and then go ahead and find the front and back 1s
+	_telem_bitmask = telem_bitmask;
+	FindFirstAndLastTelemetryPositions();
+}
+
+void VertiqTelemetryManager::FindFirstAndLastTelemetryPositions()
+{
+	uint16_t shift_val = 0x0001;
+	bool found_first = false;
+
+	//Go through from 0 to the max number of CVs we can have, and determine the first and last place we have a 1
+	for (uint8_t i = 0; i < MAX_SUPPORTABLE_IFCI_CVS; i++) {
+		if (shift_val & _telem_bitmask) {
+
+			//We only want to set the lowest value once
+			if (!found_first) {
+				_first_module_for_telem = i;
+
+				//Also initialize the current target
+				_current_telemetry_target_module_id = _first_module_for_telem;
+				found_first = true;
+			}
+
+			//Keep updating the last module for every time we hit a 1
+			_last_module_for_telem = i;
+
+			//We found another module for telemetry
+			_number_of_modules_for_telem++;
+		}
+
+		shift_val = shift_val << 1;
+	}
+}
+
+void VertiqTelemetryManager::StartPublishing(uORB::Publication<esc_status_s> *esc_status_pub)
+{
+	//Initialize our ESC status publishing
+	_esc_status.timestamp          = hrt_absolute_time();
+	_esc_status.counter            = 0;
+	_esc_status.esc_count          = _number_of_modules_for_telem;
+	_esc_status.esc_connectiontype = esc_status_s::ESC_CONNECTION_TYPE_SERIAL;
+
+	for (unsigned i = 0; i < _number_of_modules_for_telem; i++) {
+		_esc_status.esc[i].timestamp       = 0;
+		_esc_status.esc[i].esc_address     = 0;
+		_esc_status.esc[i].esc_rpm         = 0;
+		_esc_status.esc[i].esc_state       = 0;
+		_esc_status.esc[i].esc_cmdcount    = 0;
+		_esc_status.esc[i].esc_voltage     = 0;
+		_esc_status.esc[i].esc_current     = 0;
+		_esc_status.esc[i].esc_temperature = 0;
+		_esc_status.esc[i].esc_errorcount  = 0;
+		_esc_status.esc[i].failures        = 0;
+		_esc_status.esc[i].esc_power       = 0;
+	}
+
+	//Start advertising to the world
+	esc_status_pub->advertise();
+}
+
+uint16_t VertiqTelemetryManager::UpdateTelemetry()
+{
+	bool got_reply = false;
+
+	//Get the current time to check for timeout
+	hrt_abstime time_now = hrt_absolute_time();
+
+	//We timed out for this request if the time since the last request going out is greater than our timeout period
+	bool timed_out = (time_now - _time_of_last_telem_request) > _telem_timeout;
+
+	//We got a telemetry response
+	if (_motor_interface->telemetry_.IsFresh()) {
```

Comment:
if someone responded to the telemetry request, grab it, decode it, and publish it